source/common/router/upstream_request.cc

```diff
@@ -538,9 +539,19 @@ void TcpUpstream::encodeData(Buffer::Instance& data, bool end_stream) {
 }
 
 void TcpUpstream::encodeHeaders(const Http::RequestHeaderMap&, bool end_stream) {
-  if (end_stream) {
-    Buffer::OwnedImpl data;
-    upstream_conn_data_->connection().write(data, true);
+  // Headers should only happen once, so use this opportunity to add the proxy
+  // proto header, if configured.
+  ASSERT(upstream_request_->parent().routeEntry()->connectConfig().has_value());
+  Buffer::OwnedImpl data;
+  auto& connect_config = upstream_request_->parent().routeEntry()->connectConfig().value();
+  if (connect_config.has_proxy_protocol_config()) {
+    const Network::Connection& connection = *upstream_request_->parent().callbacks()->connection();
+    Extensions::Common::ProxyProtocol::generateProxyProtoHeader(
+        connect_config.proxy_protocol_config(), connection, data);
```

Comment:
We don't need to action on this in this PR, but I think this means that the proxy proto code is no longer an extension. Can we move it into the core somewhere?